File: src/main/java/it/stanislas/kafka/delay/streamjoin/ClockProducer.java
package it.stanislas.kafka.delay.streamjoin;

import io.reactivex.rxjava3.core.Observable;
import org.apache.kafka.clients.producer.KafkaProducer;
import org.apache.kafka.clients.producer.ProducerRecord;
import org.apache.kafka.clients.producer.RecordMetadata;

import java.util.concurrent.Future;
import java.util.concurrent.TimeUnit;

import static it.stanislas.kafka.delay.streamjoin.ProducerFactory.buildProducer;

public class ClockProducer {

    private final KafkaProducer<String, String> kafkaProducer;
    private final String clockTopicName;
    private final ClockKafkaKeyGenerator clockKafkaKeyGenerator;

    protected ClockProducer(final KafkaProducer<String, String> kafkaProducer,
                            final String clockTopicName,
                            final ClockKafkaKeyGenerator clockKafkaKeyGenerator
    ) {
        this.kafkaProducer = kafkaProducer;
        this.clockTopicName = clockTopicName;
        this.clockKafkaKeyGenerator = clockKafkaKeyGenerator;
    }

    public static ClockProducer buildAndStart(
            final String bootstrapServers,
            final String clockTopicName,
            final ClockKafkaKeyGenerator clockKafkaKeyGenerator
    ) {
        final ClockProducer producer = build(bootstrapServers, clockTopicName, clockKafkaKeyGenerator);
        producer.start();
        return producer;
    }

    public static ClockProducer build(
            final String bootstrapServers,
            final String clockTopicName,
            final ClockKafkaKeyGenerator clockKafkaKeyGenerator
    ) {
        return new ClockProducer(buildProducer(bootstrapServers), clockTopicName, clockKafkaKeyGenerator);
    }

    public void start() {

        Observable
                .interval(1, TimeUnit.SECONDS)
                .doOnNext(n -> {
                    sendTick();
                })
                .subscribe();

        //simulate distributed clock producer
        Observable
                .interval(1, TimeUnit.SECONDS)
                .doOnNext(n -> {
                    sendTick();
                })
                .subscribe();

    }

    private void sendTick() {
        final String key = clockKafkaKeyGenerator.now();
        final String value = key;
        final Future<RecordMetadata> recordMetadataFeature = kafkaProducer.send(new ProducerRecord(clockTopicName, key, value));
//        System.out.println("tick k:" + key + " v:" + value);
    }
}

File: src/main/kotlin/it/stanislas/kafka/delay/streamjoin/ClockProducer.kt
package it.stanislas.kafka.delay.streamjoin

import io.reactivex.rxjava3.core.Observable
import io.reactivex.rxjava3.disposables.Disposable
import org.apache.kafka.clients.producer.KafkaProducer
import org.apache.kafka.clients.producer.ProducerRecord
import java.util.concurrent.TimeUnit
import java.util.stream.IntStream


class ClockProducer (
    private val kafkaProducer: KafkaProducer<String, String>,
    private val clockTopicName: String,
    private val clockKafkaKeyGenerator: ClockKafkaKeyGenerator
) {
    private val numberOfInstances = 2
    fun start() {
        IntStream.range(0, numberOfInstances)
            .forEach { sendTickEachSecond() }
    }

    private fun sendTickEachSecond(): Disposable {
        return Observable
            .interval(1, TimeUnit.SECONDS)
            .doOnNext { sendTick() }
            .subscribe()
    }

    private fun sendTick() {
        val key = clockKafkaKeyGenerator.now()
        val recordMetadataFeature =
            kafkaProducer.send(ProducerRecord<String, String>(clockTopicName, key, key))
        //        System.out.println("tick k:" + key + " v:" + value);
    }

    companion object {
        fun buildAndStart(
            bootstrapServers: String,
            clockTopicName: String,
            clockKafkaKeyGenerator: ClockKafkaKeyGenerator
        ): ClockProducer {
            val producer =
                build(bootstrapServers, clockTopicName, clockKafkaKeyGenerator)
            producer.start()
            return producer
        }

        fun build(
            bootstrapServers: String,
            clockTopicName: String,
            clockKafkaKeyGenerator: ClockKafkaKeyGenerator
        ): ClockProducer {
            return ClockProducer(
                ProducerFactory.buildProducer(bootstrapServers),
                clockTopicName,
                clockKafkaKeyGenerator
            )
        }
    }

}